### ISSUE TITLE
Paths with trailing whitespace get mangled by the browser

### ISSUE THREAD

**Issue by nornagon** [rackt/react-router#1128]
`<a href="/foo/ ">` is treated by the browser identically to `<a href="/foo/">`, and these are equivalent:

```
history.pushState({}, "", "/test")  // leads to "/test"
history.pushState({}, "", "/test ")  // also leads to "/test"
```

So in [`Router.transitionTo`](https://github.com/rackt/react-router/blob/master/modules/createRouter.js#L247-L256), calling `location.push(<something with a trailing space>)` actually ends up pushing a different path to the one that `Router.makePath` returns.

This causes issues when there's a dynamic path component at the end of a route, and it gets set to something with a space at the end: `<Link to="foo" params={name: "bar "}>` ends up with `{name: "bar"}` in the params.

Proposed fix: call `encodeURIComponent` on every dynamic path component in `makePath`.

**Comment by santiagoaguiar** [comment]
I have the same issue, not with spaces but with other characters that need to be encoded, like /, %, &. Agree that the solution is calling `encodeURIComponent` on `injectParams`, for all params except `splat`, for which `encodeURI` should be called instead since it might actually be a path.

Any reason it's already not doing so? Does it sound ok? I'll submit a PR if so.

**Comment by mjackson** [comment]
This was fixed in #1265.

@santiagoaguiar I've got a fix on the way for `splat` which uses `encodeURI` instead of `encodeURIComponent`, as you suggested. Thank you.
